spring-data-jdbc/src/main/java/org/springframework/data/jdbc/repository/config/DefaultQueryMappingConfiguration.java

```diff
@@ -0,0 +1,53 @@
+package org.springframework.data.jdbc.repository.config;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import org.springframework.data.jdbc.repository.QueryMappingConfiguration;
+import org.springframework.jdbc.core.ResultSetExtractor;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+
+/**
+ * A {@link QueryMappingConfiguration} that allows for registration of {@link RowMapper}s and
+ * {@link ResultSetExtractor}s via a fluent Api.
+ *
+ * @author Jens Schauder
+ * @author Evgeni Dimitrov
+ */
+public class DefaultQueryMappingConfiguration implements QueryMappingConfiguration {
```

Comment:
How about splitting configuration phase from usage phase? Right now, this type is mutable which is convenient during the config phase. Turning it into immutable afterwards makes it immutable and we could add optimizations such as `AnnotationAwareOrderComparator.sort(…)` to apply ordering/overrides, especially in the case when having mappers that work with type hierarchies.